bee-tangle/src/urts.rs

```diff
@@ -0,0 +1,197 @@
+        if self.non_lazy_tips.len() > MAX_LIMIT_NON_LAZY as usize
+            || self.tips.get(parent).unwrap().children.len() as u8 > MAX_NUM_CHILDREN
+            || self
+                .tips
+                .get(parent)
+                .unwrap()
+                .time_first_child
+                .unwrap()
+                .elapsed()
+                .as_secs() as u8
+                > MAX_AGE_SECONDS_AFTER_FIRST_CHILD
+        {
+            self.tips.remove(parent);
+            self.non_lazy_tips.remove(parent);
+        }
+    }
+
+    pub(crate) async fn update_scores<B: StorageBackend>(&mut self, tangle: &MsTangle<B>) {
+        let mut to_remove = Vec::new();
+
+        for tip in self.tips.keys() {
+            match self.tip_score::<B>(tangle, &tip).await {
+                Score::SemiLazy | Score::Lazy => {
+                    to_remove.push(*tip);
+                }
+                _ => continue,
+            }
+        }
+
+        for tip in to_remove {
+            self.tips.remove(&tip);
+            self.non_lazy_tips.remove(&tip);
+        }
+
+        debug!("Non-lazy tips {}", self.non_lazy_tips.len());
+    }
+
+    async fn tip_score<B: StorageBackend>(&self, tangle: &MsTangle<B>, message_id: &MessageId) -> Score {
+        // in case the tip was pruned by the node, consider tip as lazy
+        if !tangle.contains(message_id).await {
+            return Score::Lazy;
+        }
+
+        let smi = *tangle.get_solid_milestone_index();
+        let omrsi = *tangle.omrsi(&message_id).await.unwrap().index();
+        let ymrsi = *tangle.ymrsi(&message_id).await.unwrap().index();
+
+        if (smi - ymrsi) > YMRSI_DELTA {
+            return Score::Lazy;
+        }
+
+        if (smi - omrsi) > BELOW_MAX_DEPTH {
+            return Score::Lazy;
+        }
+
+        if (smi - omrsi) > OMRSI_DELTA {
+            return Score::SemiLazy;
+        }
+
+        Score::NonLazy
+    }
+
+    pub fn two_non_lazy_tips(&self) -> Option<Vec<MessageId>> {
+        let non_lazy_tips = &self.non_lazy_tips;
+        if non_lazy_tips.is_empty() {
+            None
+        } else {
+            Some(if non_lazy_tips.len() < self.optimal_num_tips() {
+                non_lazy_tips.iter().copied().collect()
+            } else {
+                non_lazy_tips
+                    .iter()
+                    .choose_multiple(&mut rand::thread_rng(), self.optimal_num_tips())
+                    .iter()
+                    .map(|t| **t)
+                    .collect()
+            })
+        }
+    }
+
+    pub(crate) fn optimal_num_tips(&self) -> usize {
+        // TODO: hardcoded at the moment
+        4
+    }
+
+    pub(crate) fn reduce_tips(&mut self) {
+        let mut to_remove = Vec::new();
+        for (tip, metadata) in &self.tips {
+            if let Some(age) = metadata.time_first_child {
+                if age.elapsed().as_secs() as u8 > MAX_AGE_SECONDS_AFTER_FIRST_CHILD {
```

Comment:
is `u8` large enough to hold this elapsed time? or is the truncation intentional?

Reply:
you're right this must be changed, would this work?
`if age.elapsed().as_secs() > MAX_AGE_SECONDS_AFTER_FIRST_CHILD as u64`

Reply:
@rufsam Do you know what the original reason for choosing `u8` here was?

Reply:
if i remember correctly `age.elapsed().as_secs() as u8` was choosen only to be able to compare it with `MAX_AGE_SECONDS_AFTER_FIRST_CHILD` which is u8

Reply:
I've switched it to `u64` now.